### ISSUE TITLE
DBS bulk insert failed silently when data tier is missing

### ISSUE THREAD

**Issue by yuyiguo** [dmwm/DBS#498]
When insert a new dataset using insert2.py, if the data tier is missing, then nothing is inserted into the dataset table. However, if the processed ds is new, below insertion will insert the processed DS so overall, there is one row inserted and we just assumed it was the new dataset. The bulk insert/migration will failed when using the dataset id to insert dataset_output_mod_configs table because the foreign key of dataset id was missing. This needs to be fixed. It affects both migration and bulk insertion. 

insert all
           when not exists (select * from cms_dbs3_local2.processed_datasets where processed_ds_name=processed_n) then
                into cms_dbs3_local2.processed_datasets(processed_ds_id, processed_ds_name) values (cms_dbs3_local2.seq_psds.nextval, processed_n)
           when not exists (select * from cms_dbs3_local2.dataset_access_types where dataset_access_type=access_t) then
                into cms_dbs3_local2.dataset_access_types(dataset_access_type_id, dataset_access_type) values (cms_dbs3_local2.seq_dtp.nextval, access_t)
           when exists (select data_tier_id from cms_dbs3_local2.data_tiers where data_tier_name=tier) then
           into cms_dbs3_local2.datasets ( dataset_id, dataset, primary_ds_id, processed_ds_id, data_tier_id,
                           dataset_access_type_id, acquisition_era_id,  processing_era_id,
                           physics_group_id,  xtcrosssection, prep_id, creation_date, create_by,
                           last_modification_date, last_modified_by
                         )
                  values ( 280172, '/Tau/Run2015C-PromptReco-v1/MINIAOD', 250009,
                           nvl((select processed_ds_id  from cms_dbs3_local2.processed_datasets where processed_ds_name=processed_n),
                                cms_dbs3_local2.seq_psds.nextval),
                           (select data_tier_id from cms_dbs3_local2.data_tiers where data_tier_name=tier),
                          nvl((select dataset_access_type_id from cms_dbs3_local2.dataset_access_types where dataset_access_type=access_t), cms_dbs3_local2.seq_dtp.nextval),
                           260015, 65015, 100,
                           null, null, cdate, cby,
                           1439293138, 'tier0@vocms15.cern.ch'
                         )
                select  'Run2015C-PromptReco-v1' processed_n,
                        'MINIAOD' tier,  'VALID' access_t,
                        1439293138 cdate, 'tier0@vocms15.cern.ch' cby
                from dual;

**Comment by yuyiguo** [comment]
Spent quite time on this. The original thought was that the data tier can be checked during the dataset insertion, but it did not look to me Oracle SQL insert offered this kind flexible. 
So I decided to check it before the insert dataset call. Then I found that the code was already did it. but why the check of the tier had not work? The problem was that the low lever sql that check for dataset id was return -1 when not found instead of None. So the comparing with id==0(the initial value) was not satisfied because it was -1. Then we need to fake the DBS actually make the test work. Then we found there was another server connect to the same DB and migration always failed w/o any log . So cleaned everything, fixed the code and tested. Now when a missing tier, the server will report it. All done.
